src/relay/ir/hash.cc

```diff
@@ -0,0 +1,318 @@
+    //     return true;
+    //   } else {
+    //     return false;
+    //   }
+    // }
+  }
+
+  using AttrsHashHandler::VisitAttr_;
+  size_t VisitAttr_(const Variable* lhs) final {
+    return 0; // return LeafNodeEqual(GetRef<NodeRef>(lhs), other);
+  }
+
+  // Type hashing
+  size_t VisitType_(const TensorTypeNode* tensor_type) final {
+    size_t hash = std::hash<std::string>()(tensor_type->_type_key);
+    hash = Combine(hash, DataTypeHash(tensor_type->dtype));
+    hash = Combine(hash, Hash(tensor_type->shape));
+    return hash;
+  }
+
+  size_t VisitType_(const IncompleteTypeNode* incomplete) final {
+    return GetRef<IncompleteType>(incomplete);
+  }
+
+  size_t VisitType_(const TypeVarNode* tyvar) final {
+    int index = BindVar(GetRef<TypeVar>(tyvar));
+    return std::hash<int>()(index);
+  }
+
+  size_t VisitType_(const FuncTypeNode* func_type) final {
+    size_t hash = std::hash<std::string>()(func_type->_type_key);
+    for (auto type_param : func_type->type_params) {
+      hash = Combine(hash, TypeHash(type_param));
+    }
+
+    for (auto arg : func_type->arg_types) {
+      hash = Combine(hash, TypeHash(arg));
+    }
+
+    hash = Combine(hash, TypeHash(func_type->ret_type));
+    for (auto cs : func_type->type_constraints) {
+      hash = Combine(hash, TypeHash(cs));
+    }
+
+    return hash;
+  }
+
+  size_t VisitType_(const TypeRelationNode* type_rel) final {
+    size_t hash = std::hash<std::string>()(type_rel->_type_key);
+    hash = Combine(hash, std::hash<std::string>()(type_rel->func->name));
+    hash = Combine(hash, AttrHash(type_rel->attrs));
+
+    for (auto arg : type_rel->args) {
+      hash = Combine(hash, TypeHash(arg));
+    }
+
+    return hash;
+  }
+
+  size_t VisitType_(const TupleTypeNode* tuple_type) final {
+    size_t hash = std::hash<std::string>()(tuple_type->_type_key);
+    for (size_t i = 0; i < tuple_type->fields.size(); i++) {
+      hash = Combine(hash, TypeHash(tuple_type->fields[i]));
+    }
+    return hash;
+  }
+
+  // Expr equal checking.
+  size_t NDArrayHash(const runtime::NDArray& array) {
+
+    size_t hash = std::hash<uint8_t>()(array->dtype.code);
+    hash = Combine(hash, std::hash<uint8_t>()(array->dtype.bits));
+    hash = Combine(hash, std::hash<uint16_t>()(array->dtype.lanes));
+    CHECK_EQ(array->ctx.device_type, kDLCPU) << "can only compare CPU tensor";
+    size_t data_size = runtime::GetDataSize(*array.operator->());
+    uint8_t * data = reinterpret_cast<uint8_t*>(array->data);
+    for (size_t i = 0; i < data_size; i++) {
+      hash = Combine(hash, std::hash<uint8_t>()(data[i]));
+    }
+    return hash;
+  }
+
+  int BindVar(const NodeRef& var) {
+    var_map_[var] = var_counter;
+    return var_counter++;
+  }
+
+  size_t VisitExpr_(const VarNode* var) final {
+    return std::hash<int>()(var_map_[GetRef<Var>(var)]);
+  }
+
+  size_t VisitExpr_(const GlobalVarNode* global) final {
+    return GetRef<GlobalVar>(global).hash();
```

Comment:
consider hash by global->name_hint instead (so two environment might still be able to match)